docs: Add documentation for contributing additional checks to the trivy policies repo (#6234)
aquasecurity/trivy: docs/docs/scanner/misconfiguration/custom/index.md

> `# METADATA` (optional)

Reviewer: Replied [here](https://github.com/aquasecurity/trivy/pull/6234#discussion_r1511596970)

        - A `string` denoting the detected issue
            - Although `object` with `msg` field is accepted, other fields are dropped and `string` is recommended if `result.new()` is not utilised.
            - e.g. `{"msg": "deny message", "details": "something"}`

### Subtypes in the custom data

> TODO

Author: @simar7 what are subtypes?

Reviewer: Subtypes are documented [here](https://aquasecurity.github.io/trivy/v0.49/docs/scanner/misconfiguration/custom/selectors/#enabling-selectors-and-subtypes) and are only needed at this time for cloud checks.

They help us only run the checks that are needed for what's being scanned, optimizing the cloud scanning runtime a little.

    #     - type: kubernetes
    ```
  

> If you are creating checks for your Trivy misconfiguration scans, all fields are optional. The `schemas` field should be used to enable policy validation using a built-in schema. The 

Reviewer: ```suggestion
If you are creating checks for your Trivy misconfiguration scans, some fields are optional as referenced in the table below. The `schemas` field should be used to enable policy validation using a built-in schema. The 
```

correct and do not reference incorrect properties/values.

| Field name                 | Allowed values                                                    | Description                                             |         Default value         |     In table     |     In JSON      |     Required in Trivy Check     |
|----------------------------|-------------------------------------------------------------------|:-------------------------------------------------------:|:-----------------------------:|:----------------:|:----------------:| :-----------------------------: |
| title                      | Any characters                                                    | Name of the policy                                      |              N/A              | :material-check: | :material-check: |     :material-check:            |
| description                | Any characters                                                    | Description of the problem                              |              N/A              | :material-close: | :material-check: |     :material-check:            |

> | schemas.input              | `schema["kubernetes"]`, `schema["dockerfile"]`, `schema["cloud"]` | Used to validate the policy for syntax errors           |  (applied to all input types) | :material-close: | :material-close: |     :material-check:            |

Reviewer: Schema isn't required per se as it only is needed for validating the checks for functional correctness (syntax).

Furthermore, if a file has a file extension we expect scanning will still work just fine. For instance, if you try to scan argocd configs without having an argocd schema is still fine as they're treated as regular YAML files.

Reviewer: Can we format this a little better, this is how it renders for me in the browser.
<img width="398" alt="image" src="https://github.com/aquasecurity/trivy/assets/1254783/3b5f7bee-7bd8-4331-aa2a-603279757717">

| custom.service             | The service name of the provider                                  | One of the services available in the provider           |              N/A              | :material-close: | :material-check: |     :material-close:            |
| custom.short_code          | Any characters                                                    | Descriptive name for the check                          |              N/A              | :material-close: | :material-check: |     :material-check:            |
| custom.subtypes            | Cloud Resource Subtypes                                           | Refer to the section on subtypes below                  |              N/A              | :material-close: | :material-check: |     :material-close:            |
| custom.recommended_actions | Any characters                                                    | Describing what the user should do to resolve the issue |              N/A              | :material-close: | :material-check: |     :material-check:            |

> | custom.input.selector.type | Any item(s) in [this list][source-types]                          | More information provided below                         |              N/A              | :material-close: | :material-check: |     :material-check:            |

Reviewer: We need to update the link to `source-types` to point to [here](https://github.com/aquasecurity/trivy/blob/9361cdb7e28fd304d6fd2a1091feac64a6786672/pkg/iac/types/sources.go#L4)

| related_resource           | Any characters                                                    | URL to related resources                                |              N/A              | :material-close: | :material-check: |     :material-check:            |

> 

Author: @simar7  could you please check this table, specifically the "In Table" and "In JSON" parts?

Reviewer: TBH I don't understand what's the "In table" and "JSON" are for. Could you clarify?

The table is quite wide and not very easy to view as you can see below. Should we split into two table assuming we need all the columns? But nevertheless, we should move the "Required" column to the left as it's probably the most important for users to know.

<img width="920" alt="image" src="https://github.com/aquasecurity/trivy/assets/1254783/6c6f4fb5-223c-4864-9002-983c4c93c704">

Author: @knqyf263 could you please clarify what "in table" and "in json" is as we are not sure what that means? I assume the output format but not sure how it relates to writing policies?

Author: Oki, thank you

Reviewer: @AnaisUrlichs I saw the review request but I still see the JSON/table columns in the PR. Did you forget to push the change?

#### custom.avd_id and custom.id

The AVD_ID can be used to link the check to the Aqua Vulnerability Database (AVD) entry. For example, the `avd_id` `AVD-AWS-0176` is the ID of the check in the [AWS Vulnerability Database](https://avd.aquasec.com/). If you are [contributing your check to trivy-policies](../../../../community/contribute/checks/overview.md), you need to generate an ID using `make id` in the [trivy-policies](https://github.com/aquasecurity/trivy-policies) repository. The output of the command will provide you the next free IDs for the different providers in Trivy.

> The ID is based on the AVD_ID. For instanc,e if the `avd_id` is `AVD-AWS-0176`, the ID is `ID0176`.

Reviewer: ```suggestion
The ID is based on the AVD_ID. For instance if the `avd_id` is `AVD-AWS-0176`, the ID is `ID0176`.
```

#### custom.provider

> The `provider` field references the [provider](https://github.com/aquasecurity/defsec/tree/master/pkg/providers) available in Trivy. This should be the same as the provider name in the `pkg/iac/providers` directory, e.g. `aws`. 

Reviewer: Use trivy hyperlinks.

Author: I don't like them, it makes it more difficult to "read" in large documents

#### custom.service

The `service` is the name of the service by the provider that the check targets. This should be the same as the service name in the `pkg/iac/providers` directory ([Link](https://github.com/aquasecurity/trivy/tree/main/pkg/iac/providers)), e.g. `rds`.

> 

Reviewer: Services are defined within a provider, so RDS is a service and AWS is a provider. Your link seems to point to providers directory.

#### custom.input

The `input` tells Trivy what inputs this check should be applied to. Cloud provider checks should always use the `selector` input, and should always use the `type` selector with `cloud`. Check targeting Kubernetes yaml can use `kubenetes`, RBAC can use `rbac`, and so on.

#### Subtypes in the custom data

Subtypes currently only need to be defined for cloud providers [as detailed in the documentation.](./selectors.md/#enabling-selectors-and-subtypes)

> #### Scan Result

Reviewer: We inform the user about policy namespaces in [this](https://aquasecurity.github.io/trivy/v0.49/docs/scanner/misconfiguration/#pass-custom-policies) part of the docs but I wonder if we can tie that in here as well because without passing a custom namespace, the custom checks will not be evaluated (trivy by default only runs in it's own namespace, unless specified by the user).

| custom.input.selector.type | Any item(s) in [this list][source-types]                          |                              | :material-close: | :material-check: |
| url                        | Any characters                                                    |                              | :material-close: | :material-check: |

#### custom.avd_id and custom.id

> The AVD_ID can be used to link the check to the Aqua Vulnerability Database (AVD) entry. For example, the `avd_id` `AVD-AWS-0176` is the ID of the check in the [AWS Vulnerability Database](https://avd.aquasec.com/). If you are [contributing your check to trivy-policies](../../../../community/contribute/checks/overview.md), you need to generate an ID using `make id` in the [trivy-policies](https://github.com/aquasecurity/trivy-policies) repository. The output of the command will provide you the next free IDs for the different providers in Trivy.

Reviewer: ```suggestion
The AVD_ID can be used to link the check to the Aqua Vulnerability Database (AVD) entry. For example, the `avd_id` `AVD-AWS-0176` is the ID of the check in the [AWS Vulnerability Database](https://avd.aquasec.com/). If you are [contributing your check to trivy-policies](../../../../community/contribute/checks/overview.md), you need to generate an ID using `make id` in the [trivy-checks](https://github.com/aquasecurity/trivy-checks) repository. The output of the command will provide you the next free IDs for the different providers in Trivy.
```
